vars/artifactSetVersion.groovy

```diff
@@ -0,0 +1,138 @@
+            'gitCommitId',
+            'gitCredentialsId',
+            'gitUserEMail',
+            'gitUserName',
+            'gitSshUrl',
+            'tagPrefix',
+            'timestamp',
+            'timestampTemplate',
+            'versioningTemplate'
+        ]
+        Map pipelineDataMap = [
+            gitCommitId: gitUtils.getGitCommitId()
+        ]
+        List stepConfigurationKeys = [
+            'artifactType',
+            'buildTool',
+            'dockerVersionSource',
+            'filePath',
+            'gitCredentialsId',
+            'gitUserEMail',
+            'gitUserName',
+            'gitSshUrl',
+            'tagPrefix',
+            'timestamp',
+            'timestampTemplate',
+            'versioningTemplate'
+        ]
+
+        Map configuration = ConfigurationMerger.mergeWithPipelineData(parameters, parameterKeys, pipelineDataMap, stepConfiguration, stepConfigurationKeys, stepDefaults)
+
+        //buildTool is mandatory parameter
+        def utils = new Utils()
+        def buildTool = utils.getMandatoryParameter(configuration, 'buildTool')
+
+        if (!configuration.filePath)
+            configuration.filePath = configuration[buildTool].filePath //use default configuration
+
+        def newVersion
+        def artifactVersioning = ArtifactVersioning.getArtifactVersioning(buildTool, this, configuration)
+
+        if(configuration.artifactType == 'appContainer' && configuration.dockerVersionSource == 'appVersion'){
+            if (script.commonPipelineEnvironment.getArtifactVersion())
+                //replace + sign if available since + is not allowed in a Docker tag
+                newVersion = script.commonPipelineEnvironment.getArtifactVersion().replace('+', '_')
+            else
+                error ("[${stepName}] No artifact version available for 'dockerVersionSource: appVersion' -> executeBuild needs to run for the application artifact first to set the artifactVersion for the application artifact.'")
+        } else {
+            def currentVersion = artifactVersioning.getVersion()
+
+            def timestamp = configuration.timestamp ? configuration.timestamp : getTimestamp(configuration.timestampTemplate)
+
+            def versioningTemplate = configuration.versioningTemplate ? configuration.versioningTemplate : configuration[configuration.buildTool].versioningTemplate
+            //defined in default configuration
+            def binding = [version: currentVersion, timestamp: timestamp, commitId: configuration.gitCommitId]
+            def templatingEngine = new SimpleTemplateEngine()
+            def template = templatingEngine.createTemplate(versioningTemplate).make(binding)
+            newVersion = template.toString()
+        }
+
+        artifactVersioning.setVersion(newVersion)
+
+        sh 'git add .'
+
+        def gitCommitId
+
+        sshagent([configuration.gitCredentialsId]) {
+            if (configuration.gitUserName  && configuration.gitUserEMail) {
```

Comment:
user name and email might already be configured in Jenkins. Therefore this is valid.

Reply:
aahhh ...

Maybe we should check this?  
```[ -z $(git config user.email) ]```
With this is is more transparent that we take jenkins wide settings into account and my question wouln't be raised at all.  Or with other words: IMO we should try to make usage of defaults transparent.

Reply:
there are cases where the settings are not available and that is why the additional option is available.
I think we are fine with the solution to passing the settings if available or do you see an issue with this?

Reply:
>or do you see an issue with this?

It is about supportability. In case we have user.name or user.email missing and no project specific values are available we try to do the commit anyway and we will fail with the git message "[...] Please tell us who you are [...]" followed by some advises how to git config.

I can imaging a not that git experienced reader is more happy with a more specific log message telling him how to provide project specific values for user.name and user.email.

Anyhow the troubleshooter of the issue is not able to run git config is a suitable way. I can even image a less experienced trouble shooter applies the commands locally and wounder why it still does not work ... 

If you do not agree with me, keep it as it is.

Reply:
I lost you ...

What would be the snipped you suggest?

Reply:
My proposal: In case there is neither ```configuration.gitUserName``` nor ```git config user.name``` an exception should be thrown explaining that there is no git user available (similar for email) rather than executing the git commit anyway and providing the error message from ```git commit``` to the user.

Reply:
I did some tests here and this really seems to be an edge case, though it happened to a few projects before. There might be other reasons for a failing `git commit` and consequently we would  need to check for all or just leave it to the result of the git command.
I would suggest to leave it like it is for now. If we get issues in this area down the road - which I would not expect based on the experiences we have - we should act upon this.